### PR TITLE
Fix input ids can be none attn mask

### LINKED ISSUE #3289: GPT-2 attention_mask reshaping uses input_ids first dimension
# 🐛 Bug

## Information

Model I am using (Bert, XLNet ...): GPT-2

## To reproduce
Use attention_mask in **GPT2LMHead**, while feeding **inputs_embeds** instead of **input_ids**
The code fails because line 427 of modeling_gpt2.py uses input_ids first dimension to reshape the mask
```
        if attention_mask is not None:
            batch_size = input_ids.shape[0]
            attention_mask = attention_mask.view(batch_size, -1)
```
I fix it by changing **input_ids.shape[0]** to **attention_mask.shape[0]**, but I think it would be more correct to obtain single batch_size from one of the available input formats.
**Update**
   I think `batch_size = input_shape[0]` is the best way
- `transformers` version: **master**
- Python version: 3.7
- PyTorch version (GPU?):1.4

### PR DESCRIPTION
**Pull request by patrickvonplaten**
Make sure `batch_size` is correct for gpt2 and ctrl - these models need a slightly different behavior since the shape of `input_ids` can change depending on whether the past variable is inserted or not. 
See also:
PR: https://github.com/huggingface/transformers/pull/3033
and its issue:
https://github.com/huggingface/transformers/issues/3031

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Thanks for pointing out @lazarevskiVsg and @julien-c !

**Comment by patrickvonplaten** [comment]
@julien-c , will merge right away - small change!

**Comment by julien-c** [comment]
Why don’t you just use input_shape (which is always defined), to be consistent with other models?

**Comment by patrickvonplaten** [comment]
> Why don’t you just use input_shape (which is always defined), to be consistent with other models?

The problem is that GPT2 and CTRL have a different behavior when the `past` variable is inserted which previously led to problem when the attention_mask is inserted as well: 
#3031 

Therefore this slightly weird implementation.

**Comment by julien-c** [comment]
But in your code in this PR, batch_size is always input_shape[0] anyways, no?

File: src/transformers/modeling_ctrl.py [hunk 1]
```diff
@@ -330,8 +330,10 @@ def forward(
         elif input_ids is not None:
             input_shape = input_ids.size()
             input_ids = input_ids.view(-1, input_shape[-1])
```
Comment (patrickvonplaten) [inline src/transformers/modeling_ctrl.py]: The problem is this line here:
This changes the shape of `input_ids` and therefore also the `batch_size`

**Comment by patrickvonplaten** [comment]
> But in your code in this PR, batch_size is always `input_shape[0]` anyways, no?

I think in the case of CTRL and GPT2, it's actually a bigger inconsistency:

Let's say we have an input_ids tensor of shape `[batch_size, sequence_length] = [5, 4]`.

We call `GPT2Model` and save the last `output embeddings = outputs[0][:, -1, :]` **and** the `past` key/value states to speed up decoding = `outputs[1]`

Now if we want to use `past` GPT expects the `input_ids` to be of shape `[batch_size, 1] .squeezed(-1) = [batch_size]`. Therefore we have to adapt the attention mask here differently than in other models. Which is weird (and a bit suboptimal in my opinion in GPT's and CTLR's API) is that the shape of `input_ids` differs depending on whether `past` is None or not.
